src/mpi/coll/src/csel.c

```diff
@@ -0,0 +1,1254 @@
+        goto fn_exit;
+    }
+
+    int fd = open(json_file, O_RDONLY);
+    struct stat st;
+    stat(json_file, &st);
+    char *json = mmap(NULL, st.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+    close(fd);
+
+    MPIR_Csel_create_from_buf(json, create_container, csel_);
+
+  fn_exit:
+    return 0;
+}
+
+static csel_node_s *prune_tree(csel_node_s * root, MPIR_Comm * comm_ptr)
+{
+    for (csel_node_s * node = root; node;) {
+        switch (node->type) {
+            case CSEL_NODE_TYPE__OPERATOR__COMM_TYPE_INTRA:
+                if (comm_ptr->comm_kind == MPIR_COMM_KIND__INTRACOMM)
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__COMM_TYPE_INTER:
+                if (comm_ptr->comm_kind == MPIR_COMM_KIND__INTERCOMM)
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__COMM_SIZE_LE:
+                if (comm_ptr->local_size <= node->u.comm_size_le.val)
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__COMM_SIZE_LT:
+                if (comm_ptr->local_size < node->u.comm_size_lt.val)
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__COMM_SIZE_NODE_COMM_SIZE:
+                if (comm_ptr->node_comm != NULL &&
+                    MPIR_Comm_size(comm_ptr) == MPIR_Comm_size(comm_ptr->node_comm))
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__COMM_SIZE_POW2:
+                if (comm_ptr->local_size & (comm_ptr->local_size - 1))
+                    node = node->failure;
+                else
+                    node = node->success;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__COMM_SIZE_ANY:
+                node = node->success;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__IS_PARENT_COMM:
+                if (MPIR_Comm_is_parent_comm(comm_ptr) == node->u.is_parent_comm.val)
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            case CSEL_NODE_TYPE__OPERATOR__IS_NODE_CONSECUTIVE:
+                if (MPII_Comm_is_node_consecutive(comm_ptr) == node->u.is_node_consecutive.val)
+                    node = node->success;
+                else
+                    node = node->failure;
+                break;
+
+            default:
+                return node;
+        }
+    }
+    return root;
+}
+
+int MPIR_Csel_prune(void *root_csel, MPIR_Comm * comm_ptr, void **comm_csel_)
```

Comment:
Is the idea of this function to remove all redundant branches? Find an internal node in the tree that should be used for a specific communication? Something else?

It might be helpful to have a comment for this function (and at the very least any other `MPIR-` level functions so we know what they are later.

Reply:
No, it's for creating comm-specific trees.  At `MPI_INIT` time, the tree created is generic for all comms.  When a comm is created, we have more information such as comm size, intra/inter comm, etc.  So we prune the tree based on that information.

Reply:
Ok. That makes sense. I think it would be helpful to put that comment in the code since someone could attempt to read it the other way.